pkg/auth/oauth2authenticator.go

```diff
@@ -266,17 +276,29 @@ func (o *oAuth2Authenticator) AddAuthenticationHeader(request *http.Request) err
 		ctx = context.WithValue(ctx, oauth2.HTTPClient, o.httpClient)
 	}
 
-	tokenSource := o.oauthConfig.TokenSource(ctx, o.token)
-	validToken, err := tokenSource.Token()
-	if err != nil {
-		return err
-	}
+	// if the current token is invalid
+	if !o.token.Valid() {
+		globalRefreshMutex.Lock()
+		defer globalRefreshMutex.Unlock()
+
+		// check if the token in the config is invalid as well
+		token, _ := GetOAuthToken(o.config)
+		if !token.Valid() {
+			// use TokenSource to refresh the token
+			validToken, err := o.tokenRefresherFunc(ctx, o.oauthConfig, o.token)
+			if err != nil {
+				return err
+			}
 
-	if validToken != o.token {
-		o.persistToken(validToken)
+			if validToken != o.token {
+				o.persistToken(validToken)
```

Comment:
It would be good to have be able to register observers here (e.g. via callback functions), that are notified on change. That way language server could be informed of a token refresh and a new token and forward it to its Language Client.